services/github/github-commits-difference.service.js

```diff
@@ -0,0 +1,60 @@
+import Joi from 'joi'
+import { metric } from '../text-formatters.js'
+import { nonNegativeInteger } from '../validators.js'
+import { GithubAuthV3Service } from './github-auth-service.js'
+import { documentation, errorMessagesFor } from './github-helpers.js'
+
+const schema = Joi.object({ total_commits: nonNegativeInteger }).required()
+
+const queryParamSchema = Joi.object({
+  branchA: Joi.string().required(),
+  branchB: Joi.string().required(),
+}).required()
+
+export default class GithubCommitsDifference extends GithubAuthV3Service {
+  static category = 'activity'
+  static route = {
+    base: 'github/commits-difference',
+    pattern: ':user/:repo',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'GitHub commits difference between two branches/tags/commits',
+      namedParams: {
+        user: 'microsoft',
+        repo: 'vscode',
+      },
+      queryParams: {
+        branchA: '1.60.0',
+        branchB: '82f2db7',
```

Comment:
As is often the case, the hardest jobs is giving things good names :)

As demonstrated by the fact that the 2 examples you've used here are a commit hash and a tag, `branchA` and `branchB` don't actually have to be branches. Fortunately in git there is a word for this identifier that could be a branch name, commit hash or tag name, which is "ref" (you'll also notice you can use the "special" ref names just fine here like `HEAD~1`, etc if you try them).

The other thing here is that A and B aren't very descriptive either. It isn't really clear to me which one of this is expected to be the ref that is further ahead/behind. The Github API calls these params `{base}...{head}` which isn't going to be completely obvious to everyone, but I think those terms are more descriptive than A and B and at least have the advantage of maintaining consistency with the terms GitHub uses. Another option is GitLab uses `{from}...{to}` for these concepts..

So we could go with

- `base` / `head`
- `base_ref` / `head_ref`
- `from` / `to`
- `from_ref` / `to_ref`

(we use underscores in query params)

Personally, I have a weak preference for `base` and `head` just because they're the same words github uses in the API docs, but if you have a strong preference that one of the other 3 options there is more meaningful we can go with any of those. And lets make the error message "could not establish commit difference between **refs**"

Reply:
> Personally, I have a weak preference for `base` and `head` just because they're the same words github uses in the API docs, but if you have a strong preference that one of the other 3 options there is more meaningful we can go with any of those. And lets make the error message "could not establish commit difference between **refs**"

Why do we use underscores in query params (and not hyphens or camelCase)?

Sure, I'll rename `branchA` to `base`, and `branchB` to `head`.

The error message will read `"could not establish commit difference between refs"`.

Yeah, I was also not fond of putting `branch` in the URL (since users can use a couple of other options) - you're right that `ref` is a great term to encompass all of the possibilities. Thanks;)